compiler/rustc_codegen_llvm/src/common.rs

```diff
@@ -240,6 +240,13 @@ impl<'ll, 'tcx> ConstMethods<'tcx> for CodegenCx<'ll, 'tcx> {
             }
             Scalar::Ptr(ptr, _size) => {
                 let (alloc_id, offset) = ptr.into_parts();
+                // For vtables, get the underlying data allocation.
+                let alloc_id = match self.tcx.global_alloc(alloc_id) {
```

Comment:
I guess the alternative would be to move the body of the `GlobalAlloc::Memory` match arm below into a new function, and also call that from the `GlobalAlloc::Vtable` arm. Not sure which is better.

Reply:
yea, I was going to ask about that when I read the 3 copies of this snippet in the various codegen backends. Another alternative is to add a `codegen_alloc` function on `TyCtxt` that does what this snippet here does. I think I may actually prefer that instead of splitting the logic out into a function

Reply:
Hm, I think a function to codegen a `&Allocation` would be nicer. It would make sense to have a "follow the symbolic allocation" operation (which follows statics and vtables), but for statics we *don't* want to follow them in codegen, so that is just the wrong operation. Following only vtables seems rather ad-hoc.

In fact can we make the `vtable_allocation` provider return an `&Allocation` rather than an `AllocId`? And maybe not even have an `AllocId` for the "materialized" vtable? That would be best IMO.